File: app/src/main/java/app/fit/fitndflow/domain/usecase/AddCategoryUseCase.java
package app.fit.fitndflow.domain.usecase;

import android.content.Context;

import java.util.List;

import app.fit.fitndflow.data.common.SharedPrefs;
import app.fit.fitndflow.data.dto.StringInLanguagesDto;
import app.fit.fitndflow.domain.common.usecase.FitUseCase;
import app.fit.fitndflow.domain.model.CategoryModel;
import app.fit.fitndflow.domain.repository.FitnFlowRepository;
import io.reactivex.Single;

public class AddCategoryUseCase extends FitUseCase<StringInLanguagesDto, List<CategoryModel>> {
    public static final String SPANISH = "es";
    private FitnFlowRepository fitnFlowRepository;
    private Context context;

    public AddCategoryUseCase(Context context, String language, String nameCategory, FitnFlowRepository fitnFlowRepository) {
        super(convertToStringInLanguages(language, nameCategory));
        this.context = context;
        this.fitnFlowRepository = fitnFlowRepository;
    }

    @Override
    public Single<List<CategoryModel>> buildUseCaseObservable() {
        return Single.create(emitter -> {
            try {
                String apiKey = SharedPrefs.getApikeyFromSharedPRefs(context);
                List<CategoryModel> response = fitnFlowRepository.addNewCategory(inputParams, apiKey);
                emitter.onSuccess(response);
            } catch (Exception exception) {
                emitter.onError(exception);
            }
        });
    }




    //todo mover este metodo a Utils
    private static StringInLanguagesDto convertToStringInLanguages(String language, String nameCategory){
        StringInLanguagesDto stringInLanguagesDto;
        if(language.equals(SPANISH)){
            stringInLanguagesDto = new StringInLanguagesDto(nameCategory, "");
        } else {
            stringInLanguagesDto = new StringInLanguagesDto("", nameCategory);
        }
        return stringInLanguagesDto;
    }



}

File: app/src/main/java/app/fit/fitndflow/domain/usecase/AddCategoryUseCase.kt
package app.fit.fitndflow.domain.usecase

import android.content.Context
import app.fit.fitndflow.data.common.SharedPrefs
import app.fit.fitndflow.domain.common.usecase.UseCase
import app.fit.fitndflow.domain.model.CategoryModel
import app.fit.fitndflow.domain.repository.FitnFlowRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class AddCategoryUseCase(val fitnFlowRepository: FitnFlowRepository, val context: Context) : UseCase<AddCategoryUseCaseParams, List<CategoryModel>>(){
    override fun run(params: AddCategoryUseCaseParams): Flow<List<CategoryModel>> = flow {
         val apiKey = SharedPrefs.getApikeyFromSharedPRefs(context)
         val newCategory = fitnFlowRepository.addNewCategory(params.categoryName, params.language, apiKey)
        fitnFlowRepository.updateCurrentTrainingListCache(apiKey)
        emit(newCategory)
    }
}

data class AddCategoryUseCaseParams(val categoryName: String, val language: String)